oci/caps/utils.go

```diff
@@ -79,59 +85,77 @@ func inSlice(slice []string, s string) bool {
 
 // TweakCapabilities can tweak capabilities by adding or dropping capabilities
 // based on the basics capabilities.
-func TweakCapabilities(basics, adds, drops []string) ([]string, error) {
+func TweakCapabilities(basics, adds, drops []string, capabilities []string, privileged bool) ([]string, error) {
 	var (
 		newCaps []string
 		allCaps = GetAllCapabilities()
 	)
 
-	// FIXME(tonistiigi): docker format is without CAP_ prefix, oci is with prefix
-	// Currently they are mixed in here. We should do conversion in one place.
+	if privileged {
+		return allCaps, nil
+	}
+
+	if (len(adds) > 0 || len(drops) > 0) && capabilities != nil {
+		return nil, errdefs.InvalidParameter(fmt.Errorf("conflicting options: Capabilities and CapAdd / CapDrop"))
+	}
+
+	if capabilities != nil {
+		for _, cap := range capabilities {
+			if !inSlice(allCaps, cap, true) {
+				return nil, errdefs.InvalidParameter(fmt.Errorf("Unknown capability: %q", cap))
+			}
+			newCaps = append(newCaps, cap)
+		}
+		return newCaps, nil
+	}
 
 	// look for invalid cap in the drop list
 	for _, cap := range drops {
-		if strings.ToLower(cap) == "all" {
+		if strings.ToUpper(cap) == "ALL" {
 			continue
 		}
-
-		if !inSlice(allCaps, "CAP_"+cap) {
-			return nil, fmt.Errorf("Unknown capability drop: %q", cap)
+		if !strings.HasPrefix(cap, "CAP_") {
+			cap = "CAP_" + cap
+		}
+		if !inSlice(allCaps, cap, false) {
+			return nil, errdefs.InvalidParameter(fmt.Errorf("Unknown capability to drop: %q", cap))
 		}
 	}
 
 	// handle --cap-add=all
-	if inSlice(adds, "all") {
+	if inSlice(adds, "ALL", false) {
 		basics = allCaps
 	}
 
-	if !inSlice(drops, "all") {
+	if !inSlice(drops, "ALL", false) {
 		for _, cap := range basics {
 			// skip `all` already handled above
-			if strings.ToLower(cap) == "all" {
+			if strings.ToUpper(cap) == "ALL" {
 				continue
 			}
-
 			// if we don't drop `all`, add back all the non-dropped caps
-			if !inSlice(drops, cap[4:]) {
+			if !inSlice(drops, cap[4:], false) {
+				if !strings.HasPrefix(cap, "CAP_") {
+					cap = "CAP_" + cap
+				}
 				newCaps = append(newCaps, strings.ToUpper(cap))
 			}
 		}
 	}
 
 	for _, cap := range adds {
 		// skip `all` already handled above
-		if strings.ToLower(cap) == "all" {
+		if strings.ToUpper(cap) == "ALL" {
 			continue
 		}
-
-		cap = "CAP_" + cap
-
-		if !inSlice(allCaps, cap) {
-			return nil, fmt.Errorf("Unknown capability to add: %q", cap)
+		if !strings.HasPrefix(cap, "CAP_") {
```

Comment:
We should extract to a 'normalise' utility, so that we're sure we're using the same logic everywhere (and, if put in the right package, can be reused on the Client / CLI)

Reply:
Can be done but I need more clear instruction that where that should be done?